### PR TITLE
Fix possible infinite loop in pem_read_bio_key_decoder()

### LINKED ISSUE #15426: Infinite loop in decoders
Run without default provider installed.
Call PEM_read_bio_PrivateKey (or any of the PEM loading functions).

Watch pem_read_bio_key_decoder get stuck in an infinite loop.

```
52          while (!OSSL_DECODER_from_bio(dctx, bp) || pkey == NULL)
53              if (BIO_eof(bp) != 0)
54                  goto err;
```

As OSSL_DECODER_from_bio returns a zero if no encoders found.
The BIO_eof isn't going to do anything as the decoder did no work - so this code is simply incorrect in its handling logic.

```
45      int OSSL_DECODER_from_bio(OSSL_DECODER_CTX *ctx, BIO *in)
46      {
47          struct decoder_process_data_st data;
48          int ok = 0;
49          BIO *new_bio = NULL;
50
51          if (OSSL_DECODER_CTX_get_num_decoders(ctx) == 0) {
52              ERR_raise_data(ERR_LIB_OSSL_DECODER, OSSL_DECODER_R_DECODER_NOT_FOUND,
53                             "No decoders were found. For standard decoders you need "
54                             "at least one of the default or base providers "
55                             "available. Did you forget to load them?");
56              return 0;
57          }
58
59          if (BIO_tell(in) < 0) {
60              new_bio = BIO_new(BIO_f_readbuffer());
61              if (new_bio == NULL)
62                  return 0;
```

### PR DESCRIPTION
**Pull request by t8m**
There could be an infinite loop if no read happened.

Fixes #15426

### CODE REVIEW

**Review by paulidale — DISMISSED** [review]

**Comment by paulidale** [comment]
Is a test case useful for this?

**Review by slontis — COMMENTED** [review]
Please add a test case.

**Comment by t8m** [comment]
If I add a test case then without this fix it will just hang in the test. Do we want such testcase? But yeah, probably worth it anyway.

**Comment by paulidale** [comment]
I think a test is useful.  Sure it will hang but that's better than a user encountering the same thing in production.

**Comment by t8m** [comment]
> Please add a test case.

Done.

**Comment by t8m** [comment]
@paulidale The test case detected the #15442 as well, so I've fixed it too. Please re-review.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Aargh, the fips checksums seem to never be right.

File: test/evp_extra_test2.c [hunk 1]
```diff
@@ -414,6 +414,41 @@ static int test_d2i_PrivateKey_ex(int testid)
...
+    EVP_PKEY *pkey = NULL;
+
+    if (!TEST_ptr(key_bio = BIO_new_mem_buf(keydata[testid].kder, keydata[testid].size)))
+        goto err;
+    ERR_clear_error();
+    if (!TEST_ptr_null(pkey = PEM_read_bio_PrivateKey(key_bio, NULL, NULL, NULL)))
+        goto err;
+    if (!TEST_int_ne(ERR_peek_error(), 0))
+        goto err;
+    if (!TEST_ptr(provider = OSSL_PROVIDER_load(NULL, "default")))
```
Comment (slontis) [inline test/evp_extra_test2.c:431]: Should you be using a libctx here - just in case this affects other tests that could get added later.

Comment (t8m) [reply test/evp_extra_test2.c:431]: The provider is unloaded later. The NULL libctx is supposed to have only the null provider loaded so loading/unloading the default provider should not affect the other tests.

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews.

**Pull request closed without merging**